# Feature selection using GA in mGPS 

Run like: GA_feature_selection.py [crossover_min] [crossover_max] [mutation_rate] [number of offspring per generation] [initial population size] [number of reproductive units]


## 26/3 2024 

### Current architecture of GA_feature_selection.py

As of now, GA_feature_selection.py only performs feature selection to optimize the AIC of multiple linear regressions with respect to latitudes of the samples. In the future, longitudinal predictions will be optimized for as well. I am currently visualizing developing two models, one for longitude and another for latitude.
Today I have produced a first version of GA_feature_selection.py for latitude predictions. 

I have developed a python script, 'GA_feature_selection.py'. This script runs a genetic algorithm including the following steps: 
- **load_data_file(meta_file, abundance_file)**: load meta-data and abundance data
- **import_coordinates(meta_file, abundance_file)**: import coordinates from the metadata file
- **extract_predictors(df)**: extract predictor variables (all microogranisms present in the abundance file) and the meta data corresponding to the UUIDs present there from the metadata-file. 
- **extract_response_variables(df)**: extract the response-variables (longitude and latitude) 
- **initialize_population(predictors)**: A binary representation of which predictor variables will be present in the multiple linear regression to come is created. Whether a predictor is included or not is dictated by a coin flip. 
- **run_GA(population, predictors, response_variables)**: The function iterates through a number of sub-functions ((evaluate_fitness(population, predictors, response_variables), rank_populations(models), select_parents(sorted_models), crossover(parents) and mutate_offspring(offspring_population)). 
	- **evaluate_fitness(population, predictors, response_variables)**: Loops through each individual, which in turn is composed on the binary representation of genes. The microorganisms corresponding to the genes present in the individual is retrieved and a multiple-linear regression is fitted with respect to the latitude. The model number, the AIC value and the genetic make-up of the individual is returned. 
	- **rank_populations(models)**: Ranks the model with respect to AIC. The top-ranked individual has the lowest AIC, thus, informativeness of the model with respect to the number of predictors is optimized for. 
	- **select_parents(sorted_models)**: Selects the lowest and second-lowest AIC individuals to reproduce.
	- **crossover(parents)**: The genetic make-up of the lowest AIC-carrying individuals is divided at a random crossover point. Offspring are generated from their make-up.
	- **mutate_offspring(offspring_population)**: A mutation threshold is defined when running the script. Each gene in the offspring is that likely to mutate to the other binary (0 --> 1, 1 --> 0). The mutated offspring is returend.

### Feature Selection using GA_feature_selection.py, optimizing for minimal AIC when predicting latitudes of samples
GA_feature_selection.py was ran with the following conditions (python3 GA_feature_selection.py 0.3 0.9 0.2 100 100 4 50)
	- Min crossover point: 30% into the chromosomes
	- Max crossover point: 90% into the chromosomes
	- Mutation chance: 10% 
	- Offspring count: 100/generation
	- Initial population: 100 
	- Reproductive units: 4 
	- Generations: 50 

The following result was acquired: 
![The figure illustrates the lowest AIC values acquired for each generation, running GA_feature_selection.py.](https://github.com/AndrewBergman1/mGPS_GA/blob/main/fitness.png)


### Immediate actions to take: 
1) The crossover() function doesnt have a good system for mating parents. Currently, you can select how many parents you want to enter a lottery of sorts. Out of these parents, two are randomly selected. I have three options in mind, i can implement one of them: 
	- Option 1: Best parent gets to mate with random other parent.
	- Option 2: All parents above the threshold mate randomly to produce the offspring.
	- Option 3: The best mates with many, the second best mates with a few, the third best mates with fewer... etc ...

2) Figure out what the implications are of negative AIC values. 

3) Read the paper Eran sent to me.

## 27/3 2024
Today's agenda: 
- Read Eran's paper - COMPLETE
- Meet with Eran to discuss project - COMPLETE
- Think about how to improve the stochastic components of the GA: mutations, crossovers - COMPLETE
- Implement a solution for the crossover() function - COMPLETE
- Try to implement CPU paralelization in the program.
  
### Eran's Paper Take-aways
- Evaluate the GA by counting: True positives, True negatives, False negatives and False positives.

### Implementing N-point crossover for the crossover() function
There are three conventional ways of crossing over: one-point crossover, N-point crossover and uniform crossover. I will do a N-point crossover where the user can dictate the number of crossover points. This means that crossovers are randomly generated, sorted, then the parental chromosomes are alternately taken from one and the other at those crossover points. 

Current GA_feature_selection.py workflow
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/flowcharts/GA_feature_selection.py.drawio.png)

Current project plan 
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/flowcharts/mGPS_GA_workflow.drawio.png)

I ran a 200 generation long GA. The GA is seemingly still optimizing the solution, i will next try the full file for 500 generations:
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/200_gen_tests/Min%20CP:%200.3,%20Max%20CP:%200.9,%20Mut.%20Prob.%200.05,%20No.%20Offspring:%20100,%20Init%20pop%20size:%20100,%20Reproductive%20Units:%202,%20Generations:%20200,%20Crossover%20points:%203.png?raw=true)

## 28/3 2024
I ran GA_feature_selection.py for 500 generations with all data. The program has been running for 22 hours on the university server. Observing 'TIME+', it looks like most cores have only been active for a couple of minutes. This means that the majority of computations are still single-threaded. A reasonable fix for the future is further parallellization of the program. Surely, all steps can be further effectivized. 
	- Paralelize the recombination step
	- See if other parts of the code can be paralelised.
All data for 500 generations: 
The data is 3600 column x 4000 rows. On average, each individual has 2000 predictors in the multiple linear regression. This means that there are 8,000,000,000 calculations per individual per generation. As there are 100 individuals for 500 generations, this means 3,200,000,000,000,000 calculations (three quadrillion 200 trillion calculations).


## 29/3 2024
The 500 generation test failed. I got the following error message: posx and posy should be finite values, leading to a failure to plot the graph. 

I'll re-run 50 generations using the full data to see how it stacks up against the 50 generations ran using 100 observations.

The 50 generation run has the same plotting issue. 

I have now ran some short GAs on the full data [FAILED], the first 1000 entries [FAILED], the first 100 entries [SUCCESS]. I will now try the first 500 entries [SUCCESS]. First 750 failed due to 'city longitude' and 'city latitude' not being present. 

Either the data gets fucked further down the dataset or the multiple linear regression might not be able to handle the data im throwing at it. If the issue is the latter, then i think i have to apply some regularization of the data to remove collienarity. The collinearity might lead to strange values on the coefficients, which in turn leads to a failure to retrieve the AIC values. I should also try to reduce the number of possible predictor variables (just using a subset of them).
	- The fact that first_750 fails suggests that there might be data missing from the meta-data file. I should add a filter, so that if the corr. long and lat are unavailable, the observation is removed!!!
	
## 2/4 2024
The situation is as follows: 
	1. When providing 500 observations, the GA runs properly. 
	2. When providing 750 observations, the multiple linear regression is poorly fitted, leading to AIC = nan. 


Today's worK:
	1. I initially tried calculating VIF for each evualuation of fitness, this proved to be computationally intensive.
	2. I then decided to calculate VIF for all predictor variables before running the GA_feature_selection.py. The VIF calculations didnt work out because some predictor variables seem to be fully correlated. This leads to division by 0 (VIF=1/(1-R^2)). 
	3. I then decided to generate a correlation heatmap to visualise the correlation between different microorganisms. 
	4. I will decide on a correlation threshold at which i'll discard predictor variables.
	5. Calculate VIFs and handle VIFs where division by 0 occurs (catch R2=0).
		- I have generated the VIFs and collected them in a .csv file. The R2=0 still needs to be sorted out...

## 4/4 2024
I calculated the VIFs using all predictor variables and either received "inf" or nothing for each predictor variable. VIF calculations follow this equation: 1/(1-R^2). A multiple linear regression model is fitted for each predictor variable using the rest of them. In order to gain interpretable VIF values i generated a correlation matrix (pd.corr()). Using it, i removed every predictor variable that had correlations with other variables exceeding 0.7 (absolute value). I then calculated the VIFs again, now yielding interpretable values. When i was calculating VIFs using few observations, the error persisted, when calculating it with the full dataset, values were good. 

After having calculated the VIFs, i filtered away predictor varaibles with VIF > 5, as they would cause overfitting of the model. 

Followingly, I tried to run the GA. However, the AIC values returned from the multiple linear regression were still NaN. I found online that Ridge Regularisation could be used in cases where there are very many variables with relatively few observations (as in this case). Thus, i excchanged the multiple linear regression for ridge regressin. The ridge regression is based on multiple linear regression, but instead of optimising for least sum squared, it optimises for least sum squared AND (the slope)^2 * lambda (penalisation constant). In order to find the appropriate penalisation constant i employed 10-fold cross validation. I permited values from 0-39 (step=2). 

## 6/4 2024
Since i hogged the penthouse server, DAg canceled the GA. I removed the 10-fold cross validation to improve performance and makje the runs quicker. I improved the output to include R2, model parameters and all predictor variables. 

I also made a new script for making model predictions. It's not complete yet, first i need to retrieve the validation data.


## 15/4 2024
After a long run (50 generations), the predictive ability of latitude is poor (MSE = 1316). I will remove the early stopping function since it seems to halt the GA too much. 
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/results/GA_epochs/15042024.png)

Model predictive ability (OBS: R2 IS IN REALITY MSE!!!!!!!): 
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/results/model_performance/150424.png)

## 17/4 2024
Ridge regression (MSE ~ 1300) yields poor predictive ability. The predicted data is centered around lat = 0, perhaps suggesting that the coefficients are penalized to hard. Using cross-validation, alpha = 1040 (950 - 1550, step = 10). The reasoning behind using ridge regression was to deal with the significcant multicollinearity exhibited by some taxa, resulting in either VIF = None or VIF = inf. I have constructed a correlation matrix and removed one of the variables, if abs(corr) > 0.6. Then i filtered for VIF < 5. This reduces multicollinearity and should provide grounds to use MLS again.

I also realized lat_pred_mlr.py had logical issues, where the predictor variables imported from the model didn't correspond to the features that were selected for predictions. This has been resolved. 

## 18/4 2024
I have made adjustments to the prediction-script, it should now work properly. Eran spoke to me about trying to use the metasub-subset that Leo McCarthy used when developing mGPS. I have commenced a run using 100 individuals for 50 generations, we'll see how it fares. It could be a good idea to reduce the number of predictor variables in each individuals (instead of Td = 0.5, perhaps Td = 0.9). This would yield faster run times as well as fewer gits in the end. Another point of improvement would be having more crossover points.

## 20/4 2024
559415 : 3 crossover points, Td = 0.5
561511 : 10 CO, Td = 0.9

## 24/4 2024
Good results from the GA. I've ran lasso.py 0.3 0.9 0.05 500 500 2 500 3 and the following results were yielded. MSE = 122. I will try to further reduce it by implementing crossvalidation.
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/results/GA_epochs/220424.png)
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/results/model_performance/220424.png)

Although the results are a lot better than what's previously been attained, the outcome is still not good enough. longitude MSE = 1200, latitude MSE = 100. I will run a longer GA:

- python /home/andrewb/GA_feature_selection/long_GA.py 0.3 0.9 0.05 1000 1000 2 3000 5
- python /home/andrewb/GA_feature_selection/lat_GA.py 0.3 0.9 0.05 1000 1000 2 3000 5

The following results were attained: 

- GA Epochs longitude 
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/results/GA_epochs/long_model_performance_220424.png?raw=true)

- GA Epochs latitude
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/results/GA_epochs/lat_model_performance_220424.png?raw=true)

- Predictive performance longitude
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/results/model_performance/long_results_10402304.png?raw=true)

- Predictive performance latitude
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/results/model_performance/lat_results_10402304.png?raw=true)

The predictive performance is better for latitude, the reason for this i dont know. Perhaps inputing longitude and latitude as response variables in the same model would yield different results (but perhaps less accurate). I will also standardize the response- and predictor variables, which is appearently good practice when performing regressions. In parallell to experimenting with standardization, I'll make a version of the GA into a classification problem (classifying city). I will thereafter run the lat- and long-GAs.

I have now constructed a GA to classify datapoints into cities. With those predictors I will attempt to predict coordinates. So far, the classification accuracy is seemingly higher than that of Yali's RFC.

The following scripts are currently utilized: 
- city_GA.py (included in city_batch.sh): A genetic algorithm that optimizes the features that are included in a multi-logistic regression to identify cities. 
- city_pred.py: Reconstructs the multi-logistic regression used by city_GA.py and validates it on the validation data. Currently the accuracy is 91.28% (Model Accuracy: 0.9128440366972477!!!).
- long_GA.py (included in lat_batch.sh): A genetic algorithm that optimizes the features for predicting longitude. The script reads the "best_models_city" file and takes subsets of those predictors.
- lat_GA.py (included in lat_batch.sh): The same GA as above, but for latitude.

The following GA classification was ran: 
![Alt text](https://github.com/AndrewBergman1/mGPS_GA/blob/main/results/GA_epochs/GA_Classification_Performance_230424.png)

Upon testing it on validation data, the classification accuracy is 91.28%. This is higher than any classification achieved previously. Interestingly, the threshold of any feature being included in a model is 2.5%. There are 5 recombinations per offspring and the mutation rate is 10%. When initializing the population, each individual has on average 185 genes. In the end, the best model has 2290 genes (predictors). This also explains why the model runs significantly slower in the latter generations.

Leo's nvars and corr. classification accuracy:
1	50	0.869788161532979
2	100	0.878625292004757
3	200	0.882796187419915
4	300	0.87861985882159
5	500	0.874445655340271
6	1500	0.848896931934649
7	3669	0.832918807241058

Next steps:
1) Prepare to run XGBoost. Does this require finding overlap between city, country, continental, long and lat GITs?

Selected Features: Acanthamoeba polyphaga mimivirus, Acetoanaerobium sticklandii, Acetobacter senegalensis, Acetobacter tropicalis, Acetomicrobium mobile, Acholeplasma oculi, Achromobacter denitrificans, Achromobacter phage phiAxp-3, Achromobacter sp. AONIH1, Achromobacter xylosoxidans, Acidaminococcus intestini, Acidimicrobium ferrooxidans, Acidiphilium cryptum, Acidiphilium multivorum, Acidovorax citrulli, Acidovorax ebreus, Acidovorax sp. RAC01, Acidovorax sp. T1, Acinetobacter baumannii, Acinetobacter bereziniae, Acinetobacter equi, Acinetobacter johnsonii, Acinetobacter lactucae, Acinetobacter oleivorans, Acinetobacter phage Acj61, Acinetobacter phage Acj9, Acinetobacter phage IME_AB3, Acinetobacter phage LZ35, Acinetobacter phage Presley, Acinetobacter phage YMC13/03/R2096, Acinetobacter phage phiAC-1, Acinetobacter phage vB_AbaS_TRS1, Acinetobacter pittii, Acinetobacter soli, Acinetobacter sp. ADP1, Acinetobacter sp. TTH0-4, Acinetobacter virus Fri1, Acinetobacter virus PD6A3, Actinoalloteichus hoggarensis, Actinoalloteichus sp. AHMU CJ021, Actinoalloteichus sp. GBA129-24, Actinobacteria bacterium IMCC25003, Actinobacteria bacterium IMCC26077, Actinomyces meyeri, Actinomyces oris, Actinomyces pacaensis, Actinomyces radicidentis, Actinomyces sp. Chiba101, Actinomyces sp. oral taxon 414, Actinomyces virus Av1, Actinosynnema mirum, Advenella kashmirensis, Advenella mimigardefordensis, Aequorivita sublithincola, Aeribacillus pallidus, Aerococcus christensenii, Aerococcus urinae, Aeromonas caviae, Aeromonas hydrophila, Aeromonas phage CC2, Aeromonas phage phiAS5, Aeromonas sp. CA23, Aeromonas sp. O23A, Aeromonas virus 65, Aeromonas virus phiO18P, Aeropyrum pernix ovoid virus 1, Agarilytica rhodophyticola, Aggregatibacter actinomycetemcomitans, Aggregatibacter aphrophilus, Aggregatibacter phage S1249, Agrobacterium rhizogenes, Agrotis segetum nucleopolyhedrovirus B, Alcanivorax borkumensis, Alcanivorax sp. NBRC 101098, Alcanivorax xenomutans, Alkalilimnicola ehrlichii, Alphapapillomavirus 1, Alphapapillomavirus 2, Altererythrobacter ishigakiensis, Altererythrobacter marensis, Alteromonas australica, Alteromonas macleodii, Alteromonas mediterranea, Alteromonas naphthalenivorans, Alteromonas stellipolaris, Amphibacillus xylanus, Amycolatopsis orientalis, Anabaena cylindrica, Anabaena sp. 90, Anaerococcus mediterraneensis, Anaerococcus prevotii, Anaeromyxobacter sp. K, Aneurinibacillus soli, Aneurinibacillus sp. XH2, Anoxybacillus gonensis, Arcobacter nitrofigilis, Arcobacter sp. L, Arthrobacter crystallopoietes, Arthrobacter phage vB_ArS-ArV2, Arthrobacter sp. A3, Arthrobacter sp. ATCC 21022, Arthrobacter sp. IHBB 11108, Arthrobacter sp. LS16, Arthrobacter sp. QXT-31, Arthrobacter sp. YC-RL1, Arthrobacter sp. YN, Aster yellows witches'-broom phytoplasma, Asticcacaulis excentricus, Aureococcus anophagefferens virus, Autographa californica multiple nucleopolyhedrovirus, Avian endogenous retrovirus EAV-HP, Avian gyrovirus 2, Axonopus compressus streak virus, Azoarcus sp. CIB, Azoarcus sp. SY39, Azospirillum brasilense, Azospirillum humicireducens, Bacillus anthracis, Bacillus beveridgei, Bacillus cellulosilyticus, Bacillus cereus, Bacillus clausii, Bacillus endophyticus, Bacillus flexus, Bacillus gibsonii, Bacillus halodurans, Bacillus infantis, Bacillus megaterium, Bacillus muralis, Bacillus phage Mater, Bacillus phage Moonbeam, Bacillus phage vB_BanS-Tsamsa, Bacillus pseudofirmus, Bacillus pumilus, Bacillus safensis, Bacillus siamensis, Bacillus simplex, Bacillus sonorensis, Bacillus sp. ABP14, Bacillus sp. BH072, Bacillus sp. FDAARGOS_235, Bacillus sp. LM 4-2, Bacillus sp. SDLI1, Bacillus sp. YP1, Bacillus virus B103, Bacillus virus Bobb, Bacillus virus Finn, Bacillus virus Grass, Bacillus virus IEBH, Bacillus virus Pascal, Bacillus virus TP21, Bacillus virus phi29, Bacillus xiamenensis, Bacteriophage APSE-2, Bacteroides caccae, Bacteroides cellulosilyticus, Bacteroides helcogenes, Bacteroides ovatus, Bacteroides phage B40-8, Barnesiella viscericola, Bartonella apis, Bartonella clarridgeiae, Bathycoccus sp. RCC1105 virus BpV, Bernardetia litoralis, Betapapillomavirus 3, Bifidobacterium breve, Bifidobacterium catenulatum, Bifidobacterium choerinum, Bifidobacterium coryneforme, Bifidobacterium dentium, Bifidobacterium kashiwanohense, Bifidobacterium longum, Bifidobacterium pseudocatenulatum, Blastococcus saxobsidens, Blattabacterium sp. (Blaberus giganteus), Blattabacterium sp. (Blattella germanica), Blattabacterium sp. (Nauphoeta cinerea), Blautia sp. YL58, Bordetella bronchialis, Bordetella bronchiseptica, Bordetella flabilis, Bordetella genomosp. 6, Bordetella pseudohinzii, Bordetella sp. H567, Bordetella trematum, Bosea sp. AS-1, Bosea sp. PAMC 26642, Brachybacterium faecium, Brachybacterium sp. P6-10-X1, Brachybacterium sp. VM2412, Brachybacterium sp. VR2415, Brachyspira hyodysenteriae, Brachyspira intermedia, Brachyspira pilosicoli, Bradyrhizobium diazoefficiens, Bradyrhizobium icense, Bradyrhizobium japonicum, Bradyrhizobium oligotrophicum, Bradyrhizobium sp., Bradyrhizobium sp. BTAi1, Bradyrhizobium sp. ORS 278, Bradyrhizobium sp. ORS 285, Brenneria goodwinii, Brevibacillus brevis, Brevibacillus formosus, Brevibacterium linens, Brevundimonas naejangsanensis, Brevundimonas subvibrioides, Brevundimonas vesicularis, Brochothrix phage A9, Brochothrix phage BL3, Brucella melitensis, Brucella sp. 09RB8471, Brucella sp. 09RB8910, Brucella suis, Burkholderia gladioli, Burkholderia glumae, Burkholderia metallica, Burkholderia multivorans, Burkholderia phage KL3, Burkholderia phage KS14, Burkholderia plantarii, Burkholderia pyrrocinia, Burkholderia sp. CCGE1001, Burkholderia sp. CCGE1002, Burkholderia sp. MSMB0852, Burkholderia sp. PAMC 28687, Burkholderia stagnalis, Burkholderia territorii, Burkholderia thailandensis, Burkholderia ubonensis, Burkholderia vietnamiensis, Burkholderia virus DC1, Burkholderiales bacterium GJ-E10, Burkholderiales bacterium YL45, Butyrivibrio hungatei, Caldicellulosiruptor saccharolyticus, Caldilinea aerophila, Caldimicrobium thiodismutans, Calothrix sp. NIES-3974, Calothrix sp. NIES-4071, Calothrix sp. NIES-4101, Campylobacter coli, Campylobacter cuniculorum, Campylobacter curvus, Campylobacter jejuni, Campylobacter lanienae, Campylobacter sp. RM12175, Canarypox virus, Candidatus Accumulibacter phosphatis, Candidatus Baumannia cicadellinicola, Candidatus Desulforudis audaxviator, Candidatus Doolittlea endobia, Candidatus Erwinia sp. ErCipseudotaxifoliae, Candidatus Hamiltonella defensa, Candidatus Koribacter versatilis, Candidatus Kuenenia stuttgartiensis, Candidatus Methylopumilus turicensis, Candidatus Nanopelagicus limnes, Candidatus Nitrosotenuis sp. AQ6f, Candidatus Planktophila versatilis, Candidatus Portiera aleyrodidarum, Candidatus Puniceispirillum marinum, Candidatus Solibacter usitatus, Candidatus Sulcia muelleri, Candidatus Symbiobacter mobilis, Candidatus Tremblaya princeps, Canine papillomavirus 9, Capnocytophaga sp. ChDC OS43, Capnocytophaga sp. oral taxon 323, Capnocytophaga stomatis, Caulobacter henricii, Caulobacter mirabilis, Caulobacter vibrioides, Caulobacteraceae bacterium OTSz_A_272, Cedecea neteri, Celeribacter ethanolicus, Celeribacter indicus, Celeribacter manganoxidans, Cellulomonas fimi, Cellulomonas gilvus, Cellulophaga baltica, Cellulophaga lytica, Cellulosilyticum lentocellum, Cellulosimicrobium sp. TH-20, Cellvibrio sp. PSBB006, Cellvibrio sp. PSBB023, Cercopithecine betaherpesvirus 5, Chipapillomavirus 3, Chitinophaga pinensis, Chitinophagaceae bacterium 13, Chlamydia gallinacea, Chlamydia pecorum, Chloracidobacterium thermophilum, Chlorobaculum parvum, Chlorobaculum tepidum, Christensenella massiliensis, Chroococcidiopsis thermalis, Chryseobacterium gallinarum, Chryseobacterium glaciei, Chryseobacterium sp. T16E-39, Chrysochromulina ericina virus, Chrysodeixis includens nucleopolyhedrovirus, Citrobacter amalonaticus, Citrobacter freundii complex sp. CFNIH2, Citrobacter freundii complex sp. CFNIH3, Citrobacter phage Margaery, Citrobacter phage SH2, Citrobacter phage SH3, Citrobacter phage phiCFP-1, Citrobacter rodentium, Citrobacter virus Moogle, Citrobacter virus Moon, Citrus endogenous pararetrovirus, Clavibacter capsici, Clavibacter insidiosus, Clavibacter nebraskensis, Clavibacter sepedonicus, Cloacibacillus porcorum, Clostridium acetobutylicum, Clostridium argentinense, Clostridium autoethanogenum, Clostridium bornimense, Clostridium botulinum, Clostridium carboxidivorans, Clostridium formicaceticum, Clostridium kluyveri, Clostridium ljungdahlii, Clostridium novyi, Clostridium phage phiCT453B, Clostridium phage phiCT9441A, Clostridium sporogenes, Clostridium tetani, Cnuibacter physcomitrellae, Collinsella aerofaciens, Colwellia psychrerythraea, Colwellia sp. PAMC 20917, Comamonadaceae bacterium A1, Comamonadaceae bacterium B1, Conexibacter woesei, Coriobacteriaceae bacterium 68-1-3, Coriobacterium glomerans, Corynebacterium ammoniagenes, Corynebacterium argentoratense, Corynebacterium atypicum, Corynebacterium aurimucosum, Corynebacterium crudilactis, Corynebacterium doosanense, Corynebacterium epidermidicanis, Corynebacterium falsenii, Corynebacterium glaucum, Corynebacterium glyciniphilum, Corynebacterium halotolerans, Corynebacterium imitans, Corynebacterium mustelae, Corynebacterium phage P1201, Corynebacterium phocae, Corynebacterium simulans, Corynebacterium singulare, Corynebacterium sp. ATCC 6931, Corynebacterium striatum, Corynebacterium testudinoris, Corynebacterium ulcerans, Corynebacterium urealyticum, Corynebacterium ureicelerivorans, Corynebacterium uterequi, Corynebacterium variabile, Croceibacter atlanticus, Croceicoccus marinus, Cronobacter dublinensis, Cronobacter phage ENT39118, Cronobacter phage S13, Cronobacter phage vB_CsaM_GAP161, Cronobacter virus CR9, Cronobacter virus PBES02, Cryobacterium arcticum, Cryptobacterium curtum, Cupriavidus pinatubonensis, Cupriavidus sp. NH9, Cupriavidus sp. USMAA2-4, Curtobacterium sp. BH-2-1-1, Cutibacterium acnes, Cyanophage Syn30, Cyanothece sp. PCC 7822, Cyanothece sp. PCC 8801, Cyclobacterium amurskyense, Cyclobacterium marinum, Cycloclasticus sp. P1, Cycloclasticus zancles, Dactylococcopsis salina, Defluviitoga tunisiensis, Dehalobacter restrictus, Dehalobacter sp. DCA, Dehalobacterium formicoaceticum, Dehalococcoides mccartyi, Dehalogenimonas lykanthroporepellens, Deinococcus maricopensis, Deinococcus peraridilitoris, Deinococcus proteolyticus, Delftia acidovorans, Delftia phage IME-DE1, Denitrobacterium detoxificans, Denitrovibrio acetiphilus, Dermacoccus nishinomiyaensis, Desulfatibacillum alkenivorans, Desulfitobacterium hafniense, Desulfitobacterium metallireducens, Desulfobacca acetoxidans, Desulfococcus multivorans, Desulfohalobium retbaense, Desulfomicrobium baculatum, Desulfomicrobium orale, Desulfomonile tiedjei, Desulfosporosinus acidiphilus, Desulfotomaculum acetoxidans, Desulfotomaculum ferrireducens, Desulfotomaculum reducens, Desulfovibrio africanus, Desulfovibrio desulfuricans, Desulfovibrio magneticus, Desulfurispirillum indicum, Desulfuromonas sp. DDH964, Devosia sp. H5989, Devriesea agamarum, Dickeya dianthicola, Dickeya fangzhongdai, Dickeya phage RC-2014, Dickeya solani, Dinoroseobacter phage DFL12phi1, Dinoroseobacter shibae, Dokdonella koreensis, Dokdonia sp. 4H-3-7-5, Draconibacterium orientale, Dyadobacter fermentans, Dyella jiangningensis, Edwardsiella anguillarum, Edwardsiella phage PEi20, Edwardsiella sp. LADL05-105, Eikenella corrodens, Elizabethkingia genomosp. 3, Ensifer sojae, Enterobacter cloacae complex 'Hoffmann cluster IV', Enterobacter phage Arya, Enterobacter phage phiEap-2, Enterobacter sp. E20, Enterobacter sp. FY-07, Enterobacteria phage 285P, Enterobacteria phage EcoDS1, Enterobacteria phage HK225, Enterobacteria phage HK542, Enterobacteria phage J8-65, Enterobacteria phage T3, Enterobacteria phage phi80, Enterobacteriaceae bacterium ENNIH1, Enterobacteriaceae bacterium ENNIH2, Enterobacteriaceae bacterium strain FGI 57, Enterobacterial phage mEp390, Enterococcus phage EF62phi, Enterococcus phage EFC-1, Enterococcus phage EFDG1, Enterococcus phage IME-EFm5, Enterococcus phage SANTOR1, Enterococcus phage phiFL4A, Enterococcus sp. FDAARGOS_163, Enterococcus thailandicus, Enterococcus virus FL1, Enterococcus virus FL3, Enterococcus wangshanyuanii, Entomoplasma somnilux, Erwinia amylovora, Erwinia billingiae, Erwinia gerundensis, Erwinia phage vB_EamM-Y2, Erwinia phage vB_EamM_Caitlin, Erwinia phage vB_EamM_ChrisDB, Erwinia phage vB_EamM_Kwan, Erwinia phage vB_EamM_Phobos, Erwinia phage vB_EamP-S6, Erwinia pyrifoliae, Erwinia virus Ea214, Erysipelothrix larvae, Erysipelothrix rhusiopathiae, Erythrobacter flavus, Erythrobacter gangjinensis, Erythrobacter litoralis, Erythrobacter seohaensis, Escherichia albertii, Escherichia coli, Escherichia marmotae, Escherichia phage D108, Escherichia phage ECML-117, Escherichia phage Envy, Escherichia phage HK639, Escherichia phage PE3-1, Escherichia phage phAPEC8, Escherichia phage vB_EcoM_Alf5, Escherichia phage vB_EcoP_GA2A, Escherichia virus 4MG, Escherichia virus AYO145A, Escherichia virus Bp4, Escherichia virus EPS7, Escherichia virus FV3, Escherichia virus IME11, Escherichia virus JES2013, Escherichia virus JenK1, Escherichia virus K1G, Escherichia virus N15, Escherichia virus NJ01, Escherichia virus SU10, Escherichia virus Septima11, Escherichia virus wV8, Ethanoligenens harbinense, Exiguobacterium sp. MH3, Ezakiella massiliensis, Faecalibacterium prausnitzii, Ferrimonas balearica, Fervidobacterium islandicum, Fervidobacterium nodosum, Fervidobacterium pennivorans, Fibrella sp. ES10-3-2-2, Fictibacillus phosphorivorans, Filifactor alocis, Fischerella sp. NIES-3754, Flammeovirga sp. MY04, Flavivirga eckloniae, Flavobacteriaceae bacterium 3519-10, Flavobacterium columnare, Flavobacterium commune, Flavobacterium gilvum, Flavobacterium indicum, Flavobacterium johnsoniae, Formosa sp. Hel3_A1_48, Fowl aviadenovirus B, Fowl aviadenovirus E, Francisella sp. CA97-1460, Francisella sp. MA067296, Frankia casuarinae, Frankia inefficax, Frankia sp. EAN1pec, Frankia symbiont of Datisca glomerata, Frateuria aurantia, Fujinami sarcoma virus, Fusobacterium hwasookii, Fusobacterium nucleatum, Fusobacterium varium, Gammapapillomavirus 1, Gammapapillomavirus 13, Gammapapillomavirus 15, Gammapapillomavirus 18, Gammapapillomavirus 2, Gammapapillomavirus 21, Gammapapillomavirus 4, Gammapapillomavirus 5, Gammapapillomavirus 7, Geminocystis sp. NIES-3708, Geminocystis sp. NIES-3709, Gemmatimonas aurantiaca, Gemmatimonas phototrophica, Gemmatirosa kalamazoonesis, Geoalkalibacter subterraneus, Geobacillus lituanicus, Geobacillus sp. GHH01, Geobacillus thermocatenulatus, Geobacillus thermodenitrificans, Geobacillus thermoleovorans, Geobacillus virus E2, Geobacillus virus E3, Geobacter metallireducens, Geobacter pickeringii, Geobacter sp. M21, Gibbsiella quercinecans, Glaciecola nitratireducens, Glaciecola sp. 4H-3-7+YE-5, Glaesserella parasuis, Gloeobacter kilaueensis, Gluconobacter albidus, Gluconobacter oxydans, Gordonia phage BaxterFox, Gordonia phage Blueberry, Gordonia phage GRU1, Gordonia phage GTE5, Gordonia phage GTE6, Gordonia phage Lucky10, Gordonia phage Phinally, Gordonia polyisoprenivorans, Gordonia sp. YC-JH1, Gordonia terrae, Gordonia virus Kvothe, Gramella sp. LPB0144, Granulibacter bethesdensis, Gynuella sunshinyii, Haemophilus virus HP1, Hahella chejuensis, Halalkalicoccus jeotgali, Halanaerobium praevalens, Haliangium ochraceum, Halioglobus japonicus, Haliscomenobacter hydrossis, Haloarcula sp. CBA1115, Haloarcula taiwanensis, Halobacillus halophilus, Halobacillus mangrovi, Halobacterium hubeiense, Halobacteroides halobius, Halobiforma lacisalsi, Halocynthiibacter arcticus, Haloferax volcanii, Halogeometricum borinquense, Halomonas chromatireducens, Halomonas huangheensis, Halomonas hydrothermalis, Halomonas sp. 1513, Halopiger xanaduensis, Haloquadratum walsbyi, Halorubrum lacusprofundi, Halostagnicola larsenii, Haloterrigena daqingensis, Halothiobacillus sp. LS2, Hamiltonella virus APSE1, Helicobacter pylori, Herbaspirillum seropedicae, Herbinix luporum, Hirschia baltica, Human alphaherpesvirus 1, Human alphaherpesvirus 2, Human alphaherpesvirus 3, Human associated gemykibivirus 5, Human betaherpesvirus 7, Human endogenous retrovirus K, Human immunodeficiency virus 1, Human mastadenovirus B, Human mastadenovirus C, Human mastadenovirus F, Human polyomavirus 4, Human polyomavirus 5, Human polyomavirus 7, Hydrogenophaga sp. PBC, Hymenobacter sp. PAMC 26628, Hymenobacter swuensis, Hyphomicrobium denitrificans, Hyphomonas neptunium, Hyphomonas sp. Mor2, Idiomarina piscisalsi, Ignavibacterium album, Immundisolibacter cernigliae, Intrasporangium calvum, Isoptericola dokdonensis, Janibacter indicus, Janthinobacterium sp. LM6, Janthinobacterium svalbardensis, Jeongeupia sp. USM3, Jeotgalibacillus malaysiensis, Ketogulonicigenium robustum, Ketogulonicigenium vulgare, Kibdelosporangium phytohabitans, Klebsiella aerogenes, Klebsiella phage JD001, Klebsiella phage K64-1, Klebsiella phage vB_KpnM_KpV477, Klebsiella quasivariicola, Klebsiella sp. M5al, Klebsiella virus 1513, Klebsiella virus F19, Klebsiella virus KLPN1, Klebsiella virus KP15, Klebsiella virus KP27, Klebsiella virus KP34, Klebsiella virus Kp2, Klebsiella virus KpV41, Klebsiella virus KpV475, Klebsiella virus PKO111, Kluyvera intermedia, Komagataeibacter europaeus, Komagataeibacter medellinensis, Kosakonia cowanii, Kosakonia oryzae, Kurthia sp. 11kri321, Kutzneria albida, Kyrpidia sp. EA-1, Kytococcus sedentarius, Laceyella sp. FBKL4.010, Lachnoclostridium phytofermentans, Lactobacillus acetotolerans, Lactobacillus acidophilus, Lactobacillus allii, Lactobacillus crustorum, Lactobacillus curvatus, Lactobacillus delbrueckii, Lactobacillus gasseri, Lactobacillus hokkaidonensis, Lactobacillus jensenii, Lactobacillus kefiranofaciens, Lactobacillus pentosus, Lactobacillus phage KC5a, Lactobacillus phage Lc-Nu, Lactobacillus phage LfeInf, Lactobacillus phage phiAQ113, Lactobacillus phage phiPYB5, Lactobacillus phage phiadh, Lactobacillus reuteri, Lactobacillus rhamnosus, Lactobacillus sakei, Lactobacillus sanfranciscensis, Lactobacillus sp. wkB8, Lactobacillus virus LLKu, Lactobacillus virus LP65, Lactobacillus virus Lb338-1, Lactobacillus virus phiJL1, Lactococcus phage 340, Lactococcus phage 63301, Lactococcus phage 98201, Lactococcus phage BM13, Lactococcus phage D4410, Lactococcus phage P087, Lactococcus phage jm3, Lactococcus phage phi7, Lactococcus phage r1t, Lactococcus virus CB13, Lactococcus virus CB14, Lactococcus virus P008, Lawsonia intracellularis, Leclercia adecarboxylata, Leclercia sp. LSNIH1, Legionella clemsonensis, Legionella longbeachae, Legionella pneumophila, Leishmania major, Lentibacillus amyloliquefaciens, Leptolyngbya sp. NIES-3755, Leptolyngbya sp. O-77, Leptolyngbya sp. PCC 7376, Leptospira biflexa, Leptotrichia sp. oral taxon 212, Leuconostoc kimchii, Leuconostoc lactis, Listeria innocua, Listeria seeligeri, Listeria welshimeri, Listonella phage phiHSIC, Loktanella vestfoldensis, Luteimonas sp. 100111, Luteipulveratus mongoliensis, Lutibacter sp. LPB0138, Lymantria xylina nucleopolyhedrovirus, Lysobacter gummosus, Macrococcus canis, Mageeibacillus indolicus, Magnetospira sp. QH-2, Mamestra configurata nucleopolyhedrovirus A, Mannheimia haemolytica, Maribacter cobaltidurans, Maribacter sp. 1_2014MBL_MicDiv, Maricaulis maris, Marinithermus hydrothermalis, Marinobacter adhaerens, Marinobacter salarius, Marinobacter salinus, Marinobacter similis, Marinobacter sp. BSs20148, Marinobacterium aestuarii, Marinomonas mediterranea, Marinomonas sp. MWYL1, Marinovum algicola, Mariprofundus ferrinatatus, Marivivens sp. JLT3646, Martelella mediterranea, Massilia putida, Megamonas hypermegale, Meiothermus ruber, Meiothermus silvanus, Mesorhizobium amorphae, Mesorhizobium australicum, Mesorhizobium ciceri, Mesorhizobium loti, Mesorhizobium sp. WSM1497, Mesotoga prima, Methanobacterium formicicum, Methanobacterium lacus, Methanobrevibacter olleyae, Methanocella arvoryzae, Methanococcus aeolicus, Methanococcus voltae, Methanoregula boonei, Methanosaeta harundinacea, Methanosarcina mazei, Methanosarcina siciliae, Methanosarcina thermophila, Methanosarcina vacuolata, Methanothermobacter sp. CaT2, Methanothermobacter wolfeii, Methylibium petroleiphilum, Methylobacterium nodulans, Methylobacterium populi, Methylobacterium sp. C1, Methylobacterium zatmanii, Methylocystis bryophila, Methylomonas denitrificans, Methylomonas koyamae, Methylophaga frappieri, Methylophaga nitratireducenticrescens, Methylophilus sp. TWE2, Methylotenera mobilis, Methylovorus glucosotrophus, Micavibrio aeruginosavorus, Microbacterium hominis, Microbacterium phage Min1, Microbacterium sp. No. 7, Microbacterium sp. TPU 3598, Microbacterium sp. XT11, Microbacterium testaceum, Microcella alkaliphila, Micrococcus luteus, Microcoleus sp. PCC 7113, Microlunatus phosphovorus, Micromonas pusilla virus 12T, Moorea producens, Moraxella ovis, Mucilaginibacter sp. BJC16-A31, Mucilaginibacter sp. PAMC 26640, Mupapillomavirus 1, Mupapillomavirus 2, Murdochiella vaginalis, Muricauda ruestringensis, Murine leukemia virus, Mycobacterium canettii, Mycobacterium chelonae, Mycobacterium chimaera, Mycobacterium chubuense, Mycobacterium gilvum, Mycobacterium goodii, Mycobacterium haemophilum, Mycobacterium immunogenum, Mycobacterium indicus pranii, Mycobacterium leprae, Mycobacterium liflandii, Mycobacterium litorale, Mycobacterium marinum, Mycobacterium phage Tonenili, Mycobacterium sp. KMS, Mycobacterium sp. WY10, Mycobacterium sp. YC-RL4, Mycobacterium thermoresistibile, Mycobacterium vaccae, Mycobacterium virus Predator, Mycoplasma arginini, Mycoplasma canadense, Mycoplasma canis, Mycoplasma fermentans, Mycoplasma gallisepticum, Mycoplasma genitalium, Mycoplasma hominis, Mycoplasma hyopneumoniae, Mycoplasma pneumoniae, Mycoplasma pulmonis, Mycoplasma synoviae, Myxococcus fulvus, Myxococcus stipitatus, Natrialba magadii, Natrialbaceae archaeon JW/NM-HA 15, Natrinema pellirubrum, Natrinema sp. J7-2, Natronomonas moolapensis, Negativicoccus massiliensis, Neisseria gonorrhoeae, Neisseria lactamica, Neisseria sp. 10022, Neisseria zoodegmatis, Neoasaia chiangmaiensis, Nitratifractor salsuginis, Nitrosococcus halophilus, Nitrosococcus watsonii, Nitrosomonas communis, Nitrosomonas europaea, Nitrosomonas eutropha, Nitrosomonas sp. AL212, Nitrososphaera viennensis, Nitrosospira briensis, Nitrosospira lacus, Nitrosospira multiformis, Nitrospira defluvii, Nitrospira japonica, Nocardia cyriacigeorgica, Nocardia seriolae, Nocardia terpenica, Nocardioides sp. 78, Nocardiopsis dassonvillei, Nonomuraea sp. ATCC 55076, Nostoc sp. NIES-3756, Noumeavirus, Novosphingobium resinovorum, Obesumbacterium proteus, Oblitimonas alkaliphila, Oceanicoccus sagamiensis, Oceanisphaera sp. AMac2203, Oceanithermus profundus, Ochrobactrum pituitosum, Ochrobactrum pseudogrignonense, Octadecabacter antarcticus, Oenococcus oeni, Oleiphilus messinensis, Oleispira antarctica, Olsenella sp. oral taxon 807, Orgyia pseudotsugata multiple nucleopolyhedrovirus, Ornithobacterium rhinotracheale, Oscillatoria acuminata, Ostreococcus tauri virus 1, Owenweeksia hongkongensis, Oxalobacter formigenes, Paenarthrobacter aurescens, Paenibacillus beijingensis, Paenibacillus borealis, Paenibacillus bovis, Paenibacillus kribbensis, Paenibacillus phage PG1, Paenibacillus sp. FSL R5-0912, Paenibacillus sp. FSL R7-0273, Paenibacillus sp. FSL R7-0331, Paenibacillus sp. Y412MC10, Paenibacillus sp. lzh-N1, Paenibacillus terrae, Paenibacillus yonginensis, Pandoraea apista, Pandoraea norimbergensis, Pandoraea oxalativorans, Pandoraea pulmonicola, Pandoraea sputorum, Pandoraea thiooxydans, Pandoravirus dulcis, Pantoea agglomerans, Pantoea rwandensis, Pantoea sp. At-9b, Pantoea stewartii, Pantoea vagans, Parabacteroides distasonis, Parabacteroides sp. CT06, Parabacteroides sp. YL27, Paraburkholderia caribensis, Paraburkholderia hospita, Paraburkholderia phymatum, Paraburkholderia rhizoxinica, Paraburkholderia sp. BN5, Paraburkholderia sp. SOS3, Parachlamydia acanthamoebae, Paracoccus aminophilus, Paracoccus aminovorans, Paracoccus sp. CBA4604, Paracoccus zhejiangensis, Parasaccharibacter apium, Parvibaculum lavamentivorans, Parvularcula bermudensis, Pasteurella dagmatis, Paucibacter sp. KCTC 42545, Pectobacterium atrosepticum, Pectobacterium parmentieri, Pectobacterium phage ZF40, Pediococcus claussenii, Pediococcus damnosus, Pediococcus virus cIP1, Pedobacter heparinus, Pedobacter sp. PACM 27299, Pelagibaca abyssi, Pelagibacter phage HTVC010P, Pelobacter propionicus, Persicobacter sp. JZB09, Petrimonas mucosa, Phaeobacter gallaeciensis, Phaeobacter piscinae, Photobacterium damselae, Photobacterium gaetbulicola, Photobacterium profundum, Pimelobacter simplex, Piscirickettsia salmonis, Pithovirus sibericum, Planctomyces sp. SH-PL62, Planococcus faecalis, Planococcus halocryophilus, Planococcus maritimus, Planococcus plakortidis, Planococcus sp. L10.15, Planococcus sp. PAMC 21323, Plantactinospora sp. KBS50, Plautia stali symbiont, Polaribacter sp. BM10, Polaribacter sp. MED152, Polaribacter vadi, Polaromonas naphthalenivorans, Polaromonas sp. JS666, Polymorphum gilvum, Polynucleobacter asymbioticus, Polynucleobacter duraquae, Polynucleobacter necessarius, Pontibacter korlensis, Porcine type-C oncovirus, Porphyrobacter neustonensis, Porphyromonas gingivalis, Prauserella marina, Prevotella enoeca, Prevotella jejuni, Prevotella melaninogenica, Prochlorococcus phage P-GSP1, Prochlorococcus phage P-TIM68, Prochlorococcus sp. MIT 0604, Propionibacterium freudenreichii, Propionibacterium phage Enoki, Propionibacterium phage PA1-14, Propionibacterium phage PHL025, Propionibacterium phage PHL041, Propionibacterium phage PHL085, Propionibacterium phage PHL116, Propionibacterium phage PHL132, Propionibacterium phage PHL152, Propionibacterium phage PHL301, Propionibacterium virus Attacne, Propionibacterium virus P1.1, Propionibacterium virus P100A, Propionibacterium virus P100D, Propionibacterium virus P101A, Propionibacterium virus P105, Propionibacterium virus P144, Propionibacterium virus PAD20, Propionibacterium virus PAS50, Propionibacterium virus Pirate, Propionibacterium virus Stormborn, Proteus mirabilis, Proteus phage vB_PmiM_Pm5461, Providencia rettgeri, Pseudarthrobacter chlorophenolicus, Pseudarthrobacter phenanthrenivorans, Pseudarthrobacter sulfonivorans, Pseudoalteromonas atlantica, Pseudoalteromonas issachenkonii, Pseudoalteromonas luteoviolacea, Pseudoalteromonas phage pYD6-A, Pseudoalteromonas phenolica, Pseudoalteromonas spongiae, Pseudoalteromonas tetraodonis, Pseudoalteromonas tunicata, Pseudodesulfovibrio aespoeensis, Pseudomonas balearica, Pseudomonas corrugata, Pseudomonas entomophila, Pseudomonas frederiksbergensis, Pseudomonas fulva, Pseudomonas knackmussii, Pseudomonas koreensis, Pseudomonas mendocina, Pseudomonas monteilii, Pseudomonas parafulva, Pseudomonas phage 201phi2-1, Pseudomonas phage Andromeda, Pseudomonas phage F10, Pseudomonas phage JBD18, Pseudomonas phage JBD24, Pseudomonas phage JBD5, Pseudomonas phage Lu11, Pseudomonas phage PAJU2, Pseudomonas phage PPPL-1, Pseudomonas phage PaMx42, Pseudomonas phage Phi-S1, Pseudomonas phage VCM, Pseudomonas phage YH30, Pseudomonas phage YMC11/02/R656, Pseudomonas phage phi-2, Pseudomonas phage vB_PaeM_PAO1_Ab27, Pseudomonas phage vB_PaeM_PS24, Pseudomonas phage vB_PaeP_MAG4, Pseudomonas phage vB_PsyM_KIL1, Pseudomonas pseudoalcaligenes, Pseudomonas putida, Pseudomonas resinovorans, Pseudomonas sp. CCOS 191, Pseudomonas sp. FDAARGOS_380, Pseudomonas sp. FGI182, Pseudomonas sp. JY-Q, Pseudomonas sp. LPH1, Pseudomonas sp. M30-35, Pseudomonas sp. NC02, Pseudomonas sp. StFLB209, Pseudomonas sp. TKP, Pseudomonas syringae, Pseudomonas viridiflava, Pseudomonas virus 119X, Pseudomonas virus 73, Pseudomonas virus Ab09, Pseudomonas virus CAb1, Pseudomonas virus LIT1, Pseudomonas virus LKA1, Pseudomonas virus LPB1, Pseudomonas virus LUZ24, Pseudomonas virus MP38, Pseudomonas virus PAE1, Pseudomonas virus PAKP3, Pseudomonas virus PMG1, Pseudomonas virus PaMx11, Pseudomonas virus gh1, Pseudonocardia sp. HH130629-09, Pseudonocardia sp. HH130630-07, Pseudopedobacter saltans, Pseudopropionibacterium propionicum, Psychrobacter arcticus, Psychrobacter phage pOW20-A, Psychrobacter sp. AntiMn-1, Psychrobacter sp. G, Psychrobacter sp. P2G3, Psychroflexus torquis, Pusillimonas sp. T7-7, Ralstonia phage RSB1, Ralstonia solanacearum, Ralstonia virus RSF1, Ramlibacter tataouinensis, Raoultella ornithinolytica, Raoultella planticola, Rathayibacter tritici, Rhizobacter gummiphilus, Rhizobium phage RR1-A, Rhizobium sp. Kim5, Rhizobium sp. N324, Rhizobium sp. NT-26, Rhizobium sp. S41, Rhodobaca barguzinensis, Rhodobacter capsulatus, Rhodobacter sp. LPB0142, Rhodobacter virus RcCronus, Rhodobacteraceae bacterium QY30, Rhodococcus aetherivorans, Rhodococcus opacus, Rhodococcus phage CosmicSans, Rhodococcus phage REQ2, Rhodococcus phage REQ3, Rhodococcus phage ReqiDocB7, Rhodococcus pyridinivorans, Rhodococcus qingshengii, Rhodococcus ruber, Rhodococcus sp. H-CA8f, Rhodococcus sp. MTM3W5.2, Rhodococcus sp. PBTS 1, Rhodococcus virus RER2, Rhodocyclaceae bacterium, Rhodoferax antarcticus, Rhodoferax koreense, Rhodoferax saidenbachensis, Rhodomicrobium vannielii, Rhodopirellula baltica, Rhodoplanes sp. Z2-YC6860, Rhodospirillum centenum, Rhodospirillum rubrum, Rhodothermus phage RM378, Rickettsia canadensis, Rickettsia monacensis, Rickettsiales bacterium Ac37b, Robiginitalea biformata, Roseburia hominis, Roseiflexus castenholzii, Rothia dentocariosa, Rothia mucilaginosa, Rubrobacter radiotolerans, Rufibacter sp. DG15C, Ruminiclostridium thermocellum, Saccharomyces cerevisiae, Saccharophagus degradans, Saccharothrix espanaensis, Sagittula sp. P11, Salegentibacter sp. T436, Salinibacter ruber, Salinispora tropica, Salinivirga cyanobacteriivorans, Salmonella bongori, Salmonella enterica, Salmonella phage 21, Salmonella phage FSL SP-058, Salmonella phage SJ46, Salmonella virus HB2014, Salmonella virus SE2, Salmonella virus SH19, Salmonella virus SJ3, Salmonella virus SPN3US, Salmonella virus UAB87, Sebaldella termitidis, Selenomonas ruminantium, Selenomonas sputigena, Serratia fonticola, Sewage-associated gemycircularvirus 1, Shewanella baltica, Shewanella frigidimarina, Shewanella marisflavi, Shewanella piezotolerans, Shewanella psychrophila, Shewanella sp. FDAARGOS_354, Shewanella sp. MR-4, Shewanella sp. MR-7, Shewanella sp. W3-18-1, Shigella virus AG3, Shigella virus SfMu, Shimwellia blattae, Siansivirga zeaxanthinifaciens, Simkania negevensis, Singulisphaera acidiphila, Sinomonas atrocyanea, Sinorhizobium americanum, Slackia heliotrinireducens, Snodgrassella alvi, Sodalis phage phiSG1, Sodalis praecaptivus, Sodalis virus SO1, Solibacillus silvestris, Soybean chlorotic blotch virus, Soybean mild mottle virus, Sphaerochaeta globosa, Sphaerochaeta pleomorpha, Sphingobacterium mizutaii, Sphingobacterium sp. G1-14, Sphingobium indicum, Sphingobium sp. RAC03, Sphingobium sp. TKS, Sphingobium yanoikuyae, Sphingomonas sanxanigenens, Sphingomonas sp. JJ-A5, Sphingomonas sp. KC8, Sphingomonas sp. LM7, Sphingomonas sp. NIC1, Sphingopyxis alaskensis, Sphingopyxis macrogoltabida, Sphingopyxis sp. QXT-31, Sphingopyxis terrae, Sphingorhabdus flavimaris, Sphingorhabdus sp. YGSMI21, Spirochaeta thermophila, Spiroplasma apis, Spiroplasma cantharicola, Spiroplasma culicicola, Spiroplasma monobiae, Spirosoma aerolatum, Spirosoma rigui, Spodoptera frugiperda granulovirus, Spodoptera litura granulovirus, Spongiibacter sp. IMCC21906, Stackebrandtia nassauensis, Stanieria sp. NIES-3757, Staphylococcus epidermidis, Staphylococcus lutrae, Staphylococcus nepalensis, Staphylococcus pettenkoferi, Staphylococcus phage 2638A, Staphylococcus phage IME-SA4, Staphylococcus phage ROSA, Staphylococcus phage StB12, Staphylococcus sciuri, Staphylococcus virus 71, Staphylococcus virus MCE2014, Staphylococcus virus P108, Staphylococcus virus S25-4, Staphylococcus virus SA11, Staphylococcus warneri, Staphylococcus xylosus, Stenotrophomonas acidaminiphila, Stenotrophomonas phage vB_SmaS-DLP, Stenotrophomonas rhizophila, Stenotrophomonas sp. WZN-1, Stenotrophomonas virus PSH1, Stenotrophomonas virus SMA7, Streptobacillus moniliformis, Streptococcus acidominimus, Streptococcus dysgalactiae, Streptococcus halotolerans, Streptococcus infantarius, Streptococcus macedonicus, Streptococcus mitis, Streptococcus oralis, Streptococcus pantholopis, Streptococcus parasanguinis, Streptococcus phage APCM01, Streptococcus phage K13, Streptococcus phage PH10, Streptococcus phage Str-PAP-1, Streptococcus phage TP-J34, Streptococcus phage YMC-2011, Streptococcus phage phiARI0468-4, Streptococcus pneumoniae, Streptococcus sanguinis, Streptococcus sp. I-G2, Streptococcus virus 7201, Streptococcus virus 858, Streptococcus virus Sfi11, Streptococcus virus Sfi21, Streptomyces autolyticus, Streptomyces bingchenggensis, Streptomyces cattleya, Streptomyces coelicolor, Streptomyces cyaneogriseus, Streptomyces formicae, Streptomyces lavendulae, Streptomyces leeuwenhoekii, Streptomyces lydicus, Streptomyces niveus, Streptomyces peucetius, Streptomyces pluripotens, Streptomyces pratensis, Streptomyces puniciscabiei, Streptomyces rubrolavendulae, Streptomyces sampsonii, Streptomyces sp. 3211, Streptomyces sp. CCM_MD2014, Streptomyces sp. MOE7, Streptomyces sp. Mg1, Streptomyces sp. S8, Streptomyces sp. SirexAA-E, Streptomyces sp. TN58, Streptomyces sp. Tue 6075, Streptomyces sp. fd1-xmd, Streptomyces vietnamensis, Streptomyces violaceoruber, Streptomyces violaceusniger, Sulfitobacter pseudonitzschiae, Sulfuricurvum kujiense, Sulfurifustis variabilis, Sulfurimonas denitrificans, Sulfuritalea hydrogenivorans, Sulfurospirillum barnesii, Sulfurospirillum cavolei, Sulfurospirillum sp. SL2-1, Symbiobacterium thermophilum, Synechococcus lividus, Synechococcus phage ACG-2014d, Synechococcus phage ACG-2014e, Synechococcus phage ACG-2014h, Synechococcus phage S-CAM1, Synechococcus phage S-IOM18, Synechococcus phage S-RIM2, Synechococcus phage S-RSM4, Synechococcus phage S-SKS1, Synechococcus phage S-WAM1, Synechococcus sp. KORDI-100, Synechococcus sp. PCC 6312, Synechococcus sp. PCC 7117, Synechococcus sp. WH 7803, Syntrophus aciditrophicus, Tannerella forsythia, Tannerella sp. oral taxon HOT-286, Tatlockia micdadei, Tatumella citrea, Tenacibaculum dicentrarchi, Teredinibacter turnerae, Terribacillus aidingensis, Terriglobus saanensis, Tessaracoccus aquimaris, Tessaracoccus sp. T2.5-30, Thalassospira xiamenensis, Thalassotalea crassostreae, Thermacetogenium phaeum, Thermaerobacter marianensis, Thermoanaerobacter kivui, Thermoanaerobacter wiegelii, Thermoanaerobacterium xylanolyticum, Thermobacillus composti, Thermobispora bispora, Thermococcus litoralis, Thermocrinis ruber, Thermomicrobium roseum, Thermus aquaticus, Thermus brockianus, Thermus parvatiensis, Thermus sp. CCB_US3_UF1, Thermus thermophilus, Thioalkalivibrio paradoxus, Thioclava nitratireducens, Thiohalobacter thiocyanaticus, Tistrella mobilis, Tokyovirus A1, Treponema azotonutricium, Treponema brennaborense, Treponema pedis, Treponema sp. OMZ 838, Trichormus azollae, Trichormus variabilis, Trueperella pyogenes, Tsukamurella paurometabola, Tsukamurella phage TPA2, Tsukamurella phage TPA4, Tsukamurella virus TIN2, Tsukamurella virus TIN3, Turicibacter sp. H121, Turneriella parva, Variibacter gotjawalensis, Variovorax boronicumulans, Variovorax sp. PAMC 28711, Verrucomicrobia bacterium, Vibrio alginolyticus, Vibrio breoganii, Vibrio campbellii, Vibrio gazogenes, Vibrio jasicida, Vibrio nigripulchritudo, Vibrio parahaemolyticus, Vibrio phage PWH3a-P1, Vibrio phage VBP32, Vibrio phage VP93, Vibrio phage VvAW1, Vibrio phage pVp-1, Vibrio phage pYD38-A, Vibrio rotiferianus, Vibrio scophthalmi, Vibrio shilonii, Vibrio tapetis, Vibrio tritonius, Vibrio virus VP882, Vitreoscilla filiformis, Vulgatibacter incomptus, Weeksella virosa, Weissella cibaria, Weissella phage phiYS61, Wenyingzhuangia fucanilytica, Wenzhouxiangella marina, Wolbachia endosymbiont of Brugia malayi, Xanthobacter autotrophicus, Xanthomonas campestris, Xanthomonas fuscans, Xanthomonas gardneri, Xanthomonas oryzae, Xanthomonas phage vB_XveM_DIBBI, Xanthomonas sacchari, Xanthomonas vesicatoria, Xanthomonas virus CP1, Xenorhabdus nematophila, Xenorhabdus poinarii, Xylella phage Paz, Xylella virus Prado, Yangia sp. CCB-MM3, Yersinia aldovae, Yersinia phage phiR201, Yersinia phage vB_YenP_AP10, Yersinia phage vB_YenP_AP5, Yersinia rohdei, Yersinia similis, Yersinia sp. FDAARGOS_228, Zhongshania aliphaticivorans, Zunongwangia profunda, [Bacillus] selenitireducens, [Clostridium] bolteae, [Clostridium] cellulolyticum, [Clostridium] thermosuccinogenes, [Enterobacter] lignolyticus, [Eubacterium] eligens, [Mannheimia] succiniciproducens, [Pseudomonas] mesoacidophila, alpha proteobacterium HIMB5, beta proteobacterium CB, endosymbiont 'TC1' of Trimyema compressum, endosymbiont of Acanthamoeba sp. UWC8, endosymbiont of unidentified scaly snail isolate Monju, secondary endosymbiont of Heteropsylla cubana, uncultured crAssphage